### PR TITLE
In (strCommand == "tx"), return if AlreadyHave()

### PR DESCRIPTION
**Pull request by dgenr8**
A live DoS attack observed by several nodes in recent days involved repeated rejection of duplicate transactions.

Add a call to AlreadyHave when an unsolicited full tx is received, as was the case in the observed attack. AlreadyHave uses the recentRejects filter.

I tested this change on mainnet while the actual attack was still underway.

The recentRejects filter is cleared when the tip is updated, so nothing stops attacker from re-transmitting a load of rejectable txes after a new block, and in fact our attacker was doing this. But the duplicates are stopped between blocks and the attack could get arbitrarily heavy if multiple attacking peers were involved.

### CODE REVIEW

**Comment by sipa** [comment]
Concept ACK

**Comment by dajohi** [comment]
What are you thoughts on disallowing unsolicited tx's -- peers that send a tx without a getdata request.  That would break bitcoinj though.

**Comment by gavinandresen** [comment]
Untested ACK.

File: src/main.cpp [hunk 1]
```diff
@@ -4322,14 +4326,11 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
...
                 LogPrint("mempool", "mapOrphan overflow, removed %u tx\n", nEvicted);
-        } else {
-            // AcceptToMemoryPool() returned false, possibly because the tx is
-            // already in the mempool; if the tx isn't in the mempool that
-            // means it was rejected and we shouldn't ask for it again.
-            if (!mempool.exists(tx.GetHash())) {
-                assert(recentRejects);
-                recentRejects->insert(tx.GetHash());
-            }
+        } else
```
Comment (laanwj) [inline src/main.cpp]: Nit: can you please keep the style the same here

**Comment by laanwj** [comment]
utACK, nice catch

**Comment by dgenr8** [comment]
@laanwj done

**Comment by btcdrak** [comment]
utACK

**Comment by dcousens** [comment]
utACK

**Comment by jgarzik** [comment]
tested ACK

**Pull request merged**